File: Tps/Tp_2026/TP02/Questao05/Questao05.java
import java.io.File;
import java.io.PrintWriter;
import java.util.*;

class Data
{
    private int ano;
    private int mes;
    private int dia;
    
    /**
     * @author Gabriel Ferreira Pereira
     * @param ano mes e dia da data
     * @reason Construtor da classe Data
     */
    public Data ( int ano, int mes, int dia ) 
    {
        this.ano = ano;
        this.mes = mes;
        this.dia = dia;    
    }

    /**
     * @author Gabriel Ferreira Pereira
     * @param s String no formato YYYY-MM-DD
     * @reason Converte uma String para um objeto Data
     * @return objeto Data correspondente
     */
    public static Data parseData( String s )
    {
        Scanner sc = new Scanner(s);
        sc.useDelimiter("-");
        int ano = sc.nextInt();
        int mes = sc.nextInt();
        int dia = sc.nextInt();

        Data data = new Data ( ano, mes, dia );
        
        return data;
    }
    
    /**
     * @author Gabriel Ferreira Pereira
     * @reason Retorna a data no formato DD/MM/YYYY
     * @return String com a data formatada
     */
    public String formatarData ()
    {
        String s = String.format("%02d/%02d/%04d", dia,mes,ano);
        return s;
    }
}

class Hora
{
    private int hora;
    private int minuto;

    /**
     * @author Gabriel Ferreira Pereira
     * @param hora e minuto da hora
     * @reason Construtor da classe Hora
     */
    public Hora ( int hora, int minuto )
    {
        this.hora = hora;
        this.minuto = minuto;
    }

    /**
     * @author Gabriel Ferreira Pereira
     * @param s String no formato HH:mm
     * @reason Converte uma String para um objeto Hora
     * @return objeto Hora correspondente
     */
    public static Hora parseHora ( String s )
    {
        Scanner sc = new Scanner(s);
        sc.useDelimiter(":");
        int hora1 = sc.nextInt();
        int minuto = sc.nextInt();

        Hora hora = new Hora(hora1, minuto);

        return hora;
    }
    
    /**
     * @author Gabriel Ferreira Pereira
     * @reason Retorna a hora no formato HH:mm
     * @return String com a hora formatada
     */
    public String formatarHora ()
    {
        String s = String.format("%02d:%02d", hora, minuto);
        return s;
    }
}


class Restaurante
{
    private int id;
    private String nome;
    private String cidade;
    private int capacidade;
    private double avaliacao;
    private String[] tiposCozinha;
    private String faixaPreco;
    private Hora horarioAbertura;
    private Hora horarioFechamento;
    private Data dataAbertura;
    private boolean aberto;
    

    /**
     * @author Gabriel Ferreira Pereira
     * @param id, nome, cidade, capacidade, avaliacao, tiposCozinha, faixaPreco, horarioAbertura, 
     *          horarioFechamento, dataAbertura, aberto do restaurante
     * @reason Construtor da classe Restaurante
     */
    public Restaurante( int id, String nome, String cidade, int capacidade, double avaliacao, 
        String[] tiposCozinha, String faixaPreco, Hora horarioAbertura, Hora horarioFechamento, Data dataAbertura, boolean aberto ) 
    {
        this.id = id;
        this.nome = nome;
        this.cidade = cidade;
        this.capacidade = capacidade;
        this.avaliacao = avaliacao;
        this.tiposCozinha = tiposCozinha;
        this.faixaPreco = faixaPreco;
        this.horarioAbertura = horarioAbertura;
        this.horarioFechamento = horarioFechamento;
        this.dataAbertura = dataAbertura;
        this.aberto = aberto;
    }

    // Getters

    public int getID ()
    {
        return this.id;
    }

    public String getNome ()
    {
        return this.nome;
    }

    public String getCidade ()
    {
        return this.cidade;
    }

    public int getCapacidade ()
    {
        return this.capacidade;
    }

    public double getAvaliacao ()
    {
        return this.avaliacao;
    }

    public String[] getTiposCozinha ()
    {
        return this.tiposCozinha;
    }

    public String getFaixaPreco ()
    {
        return this.faixaPreco;
    }

    public Hora getHorarioAbertura ()
    {
        return this.horarioAbertura;
    }

    public Hora getHorarioFechamento ()
    {
        return this.horarioFechamento;
    }

    public Data getDataAbertura ()
    {
        return this.dataAbertura;
    }

    public boolean getAberto ()
    {
        return this.aberto;
    }
    
    /**
     * @author Gabriel Ferreira Pereira
     * @param String s
     * @reason Converte uma String para double manualmente
     * @return numero real
     */
    public static double parseDouble( String s )
    {
        int inteiro = 0;
        int decimal = 0;
        boolean ponto = false;

        for ( int i = 0; i < s.length(); i++ )
        {
            char c = s.charAt(i);
            
            if (c == '.')
            {
                ponto = true;
            }
            
            else if ( ponto == false) 
            {
                inteiro = inteiro * 10 + (c - '0');
            }

            else
            {
                decimal = decimal * 10 + (c - '0');
            }
        }

        return inteiro + decimal / 10.0;
    }

    /**
     * @author Gabriel Ferreira Pereira
     * @param path caminho do arquivo CSV
     * @reason Le uma linha do CSV e cria um objeto Restaurante
     * @return Restaurante criado
     */
    public static Restaurante ler( String linha ) throws Exception
    {
        Scanner sc = new Scanner(linha);
        sc.useDelimiter(",");

        int id = sc.nextInt();
        String nome = sc.next();
        String cidade = sc.next();
        int capacidade = sc.nextInt();
        double avaliacao = parseDouble(sc.next());

        String cozinhaStr = sc.next();
        Scanner scCozinha = new Scanner(cozinhaStr);
        scCozinha.useDelimiter(";");
        String tipo1 = scCozinha.next();
        String tipo2 = scCozinha.next();
        scCozinha.close();
        String[] tiposCozinha = new String[]{tipo1, tipo2};

        String faixaPreco = sc.next();

        String horario = sc.next();
        Scanner scHorario = new Scanner(horario);

        scHorario.useDelimiter("-");

        Hora horarioAbertura   = Hora.parseHora(scHorario.next());
        Hora horarioFechamento = Hora.parseHora(scHorario.next());
        scHorario.close();

        Data dataAbertura = Data.parseData(sc.next());

        String abertoStr = sc.next();
        boolean aberto;

        if (abertoStr.charAt(0) == 't')
        {
            aberto = true;
        }

        else
        {
            aberto = false;
        }

        Restaurante restaurante = new Restaurante(id, nome, cidade, capacidade, avaliacao, tiposCozinha, faixaPreco, horarioAbertura, horarioFechamento, dataAbertura, aberto);
        return restaurante;    
    }

    /**
     * @author Gabriel Ferreira Pereira
     * @reason Retorna o restaurante formatado como String
     * @return String com o restaurante formatado
     */
    public String formatar()
    {
        String tipos = "[" + tiposCozinha[0] + "," + tiposCozinha[1] + "]";
        String horario = horarioAbertura.formatarHora() + "-" + horarioFechamento.formatarHora();
        String data = dataAbertura.formatarData();
    
        return "[" + id + " ## " + nome + " ## " + cidade + " ## " + capacidade + " ## " + 
            avaliacao + " ## " + tipos + " ## " + faixaPreco + " ## " + 
            horario + " ## " + data + " ## " + aberto + "]";
    }

}

class ColecaoRestaurante 
{

    private int tamanho;
    private Restaurante[] restaurantes;

    public ColecaoRestaurante( int tamanho, Restaurante[] restaurantes ) 
    {
        this.tamanho = tamanho;
        this.restaurantes = restaurantes;
    }

    /**
     * @author Gabriel Ferreira Pereira
     * @reason Retorna a quantidade de restaurantes da colecao
     * @return this.tamanho
     */
    public int getTamanho() 
    {
        return this.tamanho;
    }

    /**
     * @author Gabriel Ferreira Pereira
     * @param Restaurante r
     * @reason Adiciona tamanho em ColecaoRestaurante
     */
    public void adicionar( Restaurante r )
    {
        this.restaurantes[this.tamanho] = r;
        this.tamanho++;
    }
    /**
     * @author Gabriel Ferreira Pereira
     * @reason Pega os restaurantes da colecao
     * @return this.restaurantes
     */
    public Restaurante[] getRestaurantes() 
    {
        return this.restaurantes;
    }

    /**
     * @author Gabriel Ferreira Pereira
     * @reason Imprime todos os restaurantes da colecao
     */
    public void imprimir() 
    {
        for ( int i = 0; i < this.tamanho; i++ ) 
        {
            System.out.println(this.restaurantes[i].formatar());
        }
    }

    /**
     * @author Gabriel Ferreira Pereira
     * @reason Le o dataset do arquivo CSV usando Restaurante.ler() e retorna a colecao
     * @return colecao
     */
    public static ColecaoRestaurante lerCsv() throws Exception
    {
        ColecaoRestaurante colecao = new ColecaoRestaurante(0, new Restaurante[550]);
        Scanner sc = new Scanner(new File("restaurante.csv"));
        
        sc.nextLine();

        sc.useDelimiter(",|\n");

        for ( int i = 0; i < 500; i++ )
        {
            colecao.adicionar(Restaurante.ler(sc.nextLine()));
        }

        return colecao;
    }
}

class Questao05
{

    /**
     * @author Gabriel Ferreira Pereira
     * @param selecionados, tamanho, nome, contadores
     * @reason funcao que faz a pesquisa Sequencial para achar o nome do restaurante de acordo com id fonecido
     * @return resp
     */
    public static boolean pesqSeq( Restaurante[] selecionados, int tamanho, String nome, int[] contadores )
    {
        boolean resp = false;

        for ( int i = 0; i < tamanho; i++ )
        {
            //System.out.println("'" + selecionados[i].getNome() + "'");
            contadores[0]++;
            if ( selecionados[i].getNome().compareTo(nome) == 0 )
            {
                resp = true;
                i = tamanho;
            }
            
            //System.out.println("Comparando: '" + selecionados[i].getNome() + "' com '" + nome + "'");

        }
        return resp;
    }

    /**
     * @author Gabriel Ferreira Pereira
     * @reason Metodo principal que busca e formata o restaurante com o ID fornecido
     *         e chama a funcao de pesquisa e exibe na tela SIM ou NAO se o nome fornecido esta na pesquisa
     */

    public static void main( String[] args ) throws Exception
    {
        Scanner sc = new Scanner(System.in);
        ColecaoRestaurante colecao = ColecaoRestaurante.lerCsv();
        Restaurante[] restaurantes = colecao.getRestaurantes();
        Restaurante[] selecionados = new Restaurante[500];

        int tamanho = 0;
        int id = 0;

        while ( ( id = sc.nextInt() ) != -1 )
        {
            for ( int i = 0; i < colecao.getTamanho(); i++ )
            {
                if ( restaurantes[i].getID() == id )
                {
                    selecionados[tamanho] = restaurantes[i];
                    tamanho++;
                }
            }
        }

        int[] comparacoes = new int[]{0};
        long inicio = System.currentTimeMillis();

        PrintWriter log = new PrintWriter("842527_sequencial.txt");

        sc.nextLine();    // pega o \n do nextInt de -1
        
        String nome;
        nome = sc.nextLine(); 

        while ( nome.length() != 3 || nome.charAt(0) != 'F' || nome.charAt(1) != 'I' || nome.charAt(2) != 'M' )
        {
            if ( pesqSeq(selecionados, tamanho, nome, comparacoes) )
            {
                System.out.println("SIM");
            }

            else
            {
                System.out.println("NAO");
            }

            nome = sc.nextLine();
        }

        long fim = System.currentTimeMillis();

        log.println("Comparacoes: " + comparacoes[0]);
        log.close();
    }
}

File: Tps/Tp_2026/TP03/Questao05/Questao05.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct Data
{
    int dia;
    int mes;
    int ano;
} Data;

typedef struct Hora
{
    int hora;
    int minuto;
} Hora;

typedef struct Restaurante
{
    int id;
    char nome[100];
    char cidade[100];
    int capacidade;
    double avaliacao;
    char tipo1[50];
    char tipo2[50];
    char faixaPreco[10];
    Hora horario_abertura;
    Hora horario_fechamento;
    Data data_abertura;
    int aberto;
} Restaurante;

typedef struct Colecao_restaurante
{
    int tamanho;
    Restaurante restaurantes[500];
} Colecao_restaurante;

/**
 * @author Gabriel Ferreira Pereira
 * @param s String no formato YYYY-MM-DD
 * @reason Converte uma String para um objeto Data
 * @return objeto Data correspondente
 */
Data parse_data( char *s )
{
    Data data;
    sscanf(s, "%d-%d-%d", &data.ano, &data.mes, &data.dia);
    return data;
}

/**
 * @author Gabriel Ferreira Pereira
 * @param data objeto Data
 * @param saida_linha para armazenar a data formatada
 * @reason Retorna a data no formato DD/MM/YYYY
 */
void formatar_data( Data *data, char *saida_linha )
{
    sprintf(saida_linha, "%02d/%02d/%04d", data->dia, data->mes, data->ano);
}

/**
 * @author Gabriel Ferreira Pereira
 * @param s String no formato hh:mm
 * @reason Converte uma String para um objeto Hora
 * @return objeto Hora correspondente
 */
Hora parse_hora( char *s )
{
    Hora hora;
    sscanf(s, "%d:%d", &hora.hora, &hora.minuto);
    return hora;
}

/**
 * @author Gabriel Ferreira Pereira
 * @param hora objeto Hora
 * @param saida_linha para armazenar a hora formatada
 * @reason Retorna a hora no formato hh:mm
 */
void formatar_hora( Hora *hora, char *saida_linha )
{
    sprintf(saida_linha, "%02d:%02d", hora->hora, hora->minuto);
}

/**
 * @author Gabriel Ferreira Pereira
 * @param linha, pos, campo
 * @reason Le um campo da linha do CSV ate encontrar virgula ou fim de linha
 * @return nova posicao na linha apos o campo lido
 */
int ler_campo( char *linha, int pos, char *campo )
{
    int i = 0;
    
    while ( linha[pos] != ',' && linha[pos] != '\0' && linha[pos] != '\n' )
    {
        campo[i] = linha[pos];
        i++;
        pos++;
    }

    campo[i] = '\0';
    pos++;

    return pos;
}

/**
 * @author Gabriel Ferreira Pereira
 * @param linha String da linha do CSV
 * @reason Le uma linha do CSV e cria um objeto Restaurante
 * @return Restaurante criado
 */
Restaurante ler_restaurante( char *linha )
{
    Restaurante r;
    char campo[100];
    char horario_abertura[10];
    char horario_fechamento[10];
    char data_str[20];
    char aberto_str[10];
    char cozinha[100];
    int pos = 0;
    int j = 0, k = 0;

    // id
    pos = ler_campo(linha, pos, campo);
    sscanf(campo, "%d", &r.id);

    // nome
    pos = ler_campo(linha, pos, r.nome);

    // cidade
    pos = ler_campo(linha, pos, r.cidade);

    // capacidade
    pos = ler_campo(linha, pos, campo);
    sscanf(campo, "%d", &r.capacidade);

    // avaliacao
    pos = ler_campo(linha, pos, campo);
    sscanf(campo, "%lf", &r.avaliacao);

    // cozinha
    pos = ler_campo(linha, pos, cozinha);

    while ( cozinha[j] != ';' )
    {
        r.tipo1[k++] = cozinha[j++];
    }

    r.tipo1[k] = '\0';
    j++; k = 0;
    
    while ( cozinha[j] != '\0' )
    {
        r.tipo2[k++] = cozinha[j++];
    }
    r.tipo2[k] = '\0';

    // faixa preco
    pos = ler_campo(linha, pos, r.faixaPreco);

    // horario
    pos = ler_campo(linha, pos, campo);
    j = 0; k = 0;

    while ( campo[j] != '-' )
    {
        horario_abertura[k++] = campo[j++];
    }

    horario_abertura[k] = '\0';
    j++; k = 0;

    while ( campo[j] != '\0' )
    {
        horario_fechamento[k++] = campo[j++];
    }

    horario_fechamento[k] = '\0';
    r.horario_abertura = parse_hora(horario_abertura);
    r.horario_fechamento = parse_hora(horario_fechamento);

    // data
    pos = ler_campo(linha, pos, data_str);
    r.data_abertura = parse_data(data_str);

    // aberto
    pos = ler_campo(linha, pos, aberto_str);

    if ( aberto_str[0] == 't' )
    {
        r.aberto = 1;
    }

    else
    {
        r.aberto = 0;
    }

    return r;
}

/**
 * @author Gabriel Ferreira Pereira
 * @param r, saida_linha
 * @reason Retorna o restaurante formatado como String
 */
void formatar_restaurante( Restaurante *r, char *saida_linha )
{
    char data_str[20];
    char horario_abertura[10];
    char horario_fechamento[10];
    char aberto_str[10];

    formatar_data(&r->data_abertura, data_str);
    formatar_hora(&r->horario_abertura, horario_abertura);
    formatar_hora(&r->horario_fechamento, horario_fechamento);

    if ( r->aberto == 1 )
    {
        sprintf(aberto_str, "true");
    }

    else
    {
        sprintf(aberto_str, "false");
    }

    sprintf(saida_linha, "[%d ## %s ## %s ## %d ## %.1f ## [%s,%s] ## %s ## %s-%s ## %s ## %s]",
        r->id, r->nome, r->cidade, r->capacidade, r->avaliacao,
        r->tipo1, r->tipo2, r->faixaPreco,
        horario_abertura, horario_fechamento,
        data_str, aberto_str);
}

/**
 * @author Gabriel Ferreira Pereira
 * @param colecao objeto Colecao_restaurante
 * @reason Retorna a quantidade de restaurantes da colecao
 * @return tamanho da colecao
 */
int get_tamanho( Colecao_restaurante *colecao )
{
    return colecao->tamanho;
}

/**
 * @author Gabriel Ferreira Pereira
 * @param colecao, r
 * @reason Adiciona um restaurante na colecao
 */
void adicionar( Colecao_restaurante *colecao, Restaurante r )
{
    colecao->restaurantes[colecao->tamanho] = r;
    colecao->tamanho++;
}

/**
 * @author Gabriel Ferreira Pereira
 * @param colecao objeto Colecao_restaurante
 * @reason Retorna o array de restaurantes da colecao
 * @return ponteiro para o array de restaurantes
 */
Restaurante* get_restaurantes( Colecao_restaurante *colecao )
{
    return colecao->restaurantes;
}

/**
 * @author Gabriel Ferreira Pereira
 * @param colecao objeto Colecao_restaurante
 * @reason Imprime todos os restaurantes da colecao formatados
 */
void imprimir( Colecao_restaurante *colecao )
{
    char saida_linha[500];
    for ( int i = 0; i < colecao->tamanho; i++ )
    {
        formatar_restaurante(&colecao->restaurantes[i], saida_linha);
        printf("%s\n", saida_linha);
    }
}

/**
 * @author Gabriel Ferreira Pereira
 * @reason Le o dataset do arquivo CSV e retorna a colecao de restaurantes
 * @return colecao de restaurantes
 */
Colecao_restaurante ler_csv()
{
    Colecao_restaurante colecao;
    colecao.tamanho = 0;

    FILE *f = fopen("/tmp/restaurantes.csv", "r");
    char linha[500];
    int j = 0;

    // pular cabecalho
    fgets(linha, 500, f);

    for ( int i = 0; i < 500; i++ )
    {
        fgets(linha, 500, f);
        j = 0;

        // substitui o \n por \0 para encerrar a string
        while ( linha[j] != '\n' && linha[j] != '\0' )
        {
            j++;
        }
        linha[j] = '\0';
        
        Restaurante r = ler_restaurante(linha);
        adicionar(&colecao, r);
    }

    fclose(f);
    return colecao;
}


typedef struct Celula 
{
	Restaurante elemento;
	struct Celula* prox;
} Celula;

Celula* primeiro;
Celula* ultimo;

/**
 * @author Gabriel Ferreira Pereira
 * @param restaurante
 * @reason Cria uma nova celula da lista flexivel
 * @return ponteiro para a nova celula
 */
Celula* novaCelula( Restaurante restaurante ) 
{
   Celula* nova = (Celula*) malloc(sizeof(Celula));
   nova->elemento = restaurante;
   nova->prox = NULL;
   return nova;
}

/**
 * @author Gabriel Ferreira Pereira
 * @reason Cria uma lista sem elementos com celula cabeca
 */
void start () 
{
   Restaurante vazio;
   vazio.id = 0;
   vazio.nome[0] = '\0';
   vazio.cidade[0] = '\0';
   vazio.capacidade = 0;
   vazio.avaliacao = 0.0;
   vazio.tipo1[0] = '\0';
   vazio.tipo2[0] = '\0';
   vazio.faixaPreco[0] = '\0';
   vazio.horario_abertura.hora = 0;
   vazio.horario_abertura.minuto = 0;
   vazio.horario_fechamento.hora = 0;
   vazio.horario_fechamento.minuto = 0;
   vazio.data_abertura.dia = 0;
   vazio.data_abertura.mes = 0;
   vazio.data_abertura.ano = 0;
   vazio.aberto = 0;

   primeiro = novaCelula(vazio);
   ultimo = primeiro;
}

/**
 * @author Gabriel Ferreira Pereira
 * @param restaurante
 * @reason Insere um restaurante na primeira posicao da lista
 */
void inserirInicio( Restaurante restaurante ) 
{
   Celula* tmp = novaCelula(restaurante);
   tmp->prox = primeiro->prox;
   primeiro->prox = tmp;

   if ( primeiro == ultimo ) 
   {                    
        ultimo = tmp;
   }
}

/**
 * @author Gabriel Ferreira Pereira
 * @param restaurante
 * @reason Insere um restaurante na ultima posicao da lista
 */
void inserirFim( Restaurante restaurante ) 
{
   ultimo->prox = novaCelula(restaurante);
   ultimo = ultimo->prox;
}

/**
 * @author Gabriel Ferreira Pereira
 * @reason Remove um restaurante da primeira posicao da lista
 * @return restaurante removido
 */
Restaurante removerInicio() 
{
   Celula* tmp = primeiro->prox;
   Restaurante resp = tmp->elemento;
   primeiro->prox = tmp->prox;

   if ( tmp == ultimo )
   {
      ultimo = primeiro;
   }

   tmp->prox = NULL;
   free(tmp);
   return resp;
}

/**
 * @author Gabriel Ferreira Pereira
 * @reason Remove um restaurante da ultima posicao da lista
 * @return restaurante removido
 */
Restaurante removerFim() 
{
   Celula* i;

   for ( i = primeiro; i->prox != ultimo; i = i->prox );

   Restaurante resp = ultimo->elemento;
   free(ultimo);
   ultimo = i;
   ultimo->prox = NULL;

   return resp;
}

/**
 * @author Gabriel Ferreira Pereira
 * @reason Calcula e retorna o tamanho da lista
 * @return tamanho da lista
 */
int tamanho() 
{
   int tamanho = 0;
   Celula* i;
   for ( i = primeiro; i != ultimo; i = i->prox, tamanho++ );
   return tamanho;
}

/**
 * @author Gabriel Ferreira Pereira
 * @param restaurante, pos
 * @reason Insere um restaurante em uma posicao especifica da lista
 */
void inserir( Restaurante restaurante, int pos ) 
{
   int tam = tamanho();

   if ( pos == 0 )
   {
      inserirInicio(restaurante);
   } 
   else if ( pos == tam )
   {
      inserirFim(restaurante);
   } 

   else 
   {
      int j;
      Celula* i = primeiro;

      for ( j = 0; j < pos; j++, i = i->prox );

      Celula* tmp = novaCelula(restaurante);
      tmp->prox = i->prox;
      i->prox = tmp;
   }
}

/**
 * @author Gabriel Ferreira Pereira
 * @param pos
 * @reason Remove um restaurante de uma posicao especifica da lista
 * @return restaurante removido
 */
Restaurante remover( int pos ) 
{
   Restaurante resp;
   int tam = tamanho();

   if ( pos == 0 )
   {
      resp = removerInicio();
   } 

   else if ( pos == tam - 1 )
   {
      resp = removerFim();
   } 

   else 
   {
      Celula* i = primeiro;
      int j;

      for ( j = 0; j < pos; j++, i = i->prox );

      Celula* tmp = i->prox;
      resp = tmp->elemento;
      i->prox = tmp->prox;
      tmp->prox = NULL;
      free(tmp);
   }

   return resp;
}

/**
 * @author Gabriel Ferreira Pereira
 * @reason Mostra os restaurantes da lista formatados
 */
void mostrar() 
{
   char saida_linha[500];
   Celula* i;
   
   for ( i = primeiro->prox; i != NULL; i = i->prox ) 
   {
      formatar_restaurante(&i->elemento, saida_linha);
      printf("%s\n", saida_linha);
   }
}

/**
 * @author Gabriel Ferreira Pereira
 * @param restaurantes, tamanho, id
 * @reason Busca um restaurante pelo ID no vetor lido do CSV
 * @return restaurante encontrado
 */
Restaurante buscar_restaurante( Restaurante *restaurantes, int tamanho, int id )
{
    for ( int i = 0; i < tamanho; i++ )
    {
        if ( restaurantes[i].id == id )
        {
            return restaurantes[i];
        }
    }

    return restaurantes[0];
}

/**
 * @author Gabriel Ferreira Pereira
 * @reason Metodo principal que busca e formata o restaurante com o ID fornecido
 *         e executa os comandos sobre a lista flexivel
 */
int main()
{
    Colecao_restaurante colecao = ler_csv();
    Restaurante *restaurantes = get_restaurantes(&colecao);
    int id = 0;
    int n = 0;
    char comando[3];

    start();

    while ( scanf("%d", &id) && id != -1 )
    {
        inserirFim(buscar_restaurante(restaurantes, get_tamanho(&colecao), id));
    }

    scanf("%d", &n);

    for ( int i = 0; i < n; i++ )
    {
        scanf("%s", comando);

        if ( comando[0] == 'I' && comando[1] == 'I' )
        {
            scanf("%d", &id);
            inserirInicio(buscar_restaurante(restaurantes, get_tamanho(&colecao), id));
        }

        else if ( comando[0] == 'I' && comando[1] == '*' )
        {
            int pos = 0;
            scanf("%d %d", &pos, &id);
            inserir(buscar_restaurante(restaurantes, get_tamanho(&colecao), id), pos);
        }

        else if ( comando[0] == 'I' && comando[1] == 'F' )
        {
            scanf("%d", &id);
            inserirFim(buscar_restaurante(restaurantes, get_tamanho(&colecao), id));
        }

        else if ( comando[0] == 'R' && comando[1] == 'I' )
        {
            Restaurante removido = removerInicio();
            printf("(R)%s\n", removido.nome);
        }

        else if ( comando[0] == 'R' && comando[1] == '*' )
        {
            int pos = 0;
            scanf("%d", &pos);
            Restaurante removido = remover(pos);
            printf("(R)%s\n", removido.nome);
        }

        else if ( comando[0] == 'R' && comando[1] == 'F' )
        {
            Restaurante removido = removerFim();
            printf("(R)%s\n", removido.nome);
        }
    }

    mostrar();

}
